src/Servers/Kestrel/Core/src/Internal/Http/HttpParser.cs

```diff
@@ -236,159 +234,234 @@ public unsafe bool ParseHeaders(TRequestHandler handler, ref SequenceReader<byte
+                        ch = headerLine[valueStart];
+                        // Do a fast check; as we now expect non-space, before moving into loop.
+                        if (ch <= ByteSpace && (ch == ByteSpace || ch == ByteTab))
+                        {
+                            valueStart++;
+                            // Is more whitespace, so we will loop to find the end. This is the slow path.
+                            for (; valueStart < headerLine.Length; valueStart++)
+                            {
+                                ch = headerLine[valueStart];
+                                if (ch != ByteTab && ch != ByteSpace)
+                                {
+                                    // Non-whitespace char found, valueStart is now start of value.
+                                    break;
+                                }
+                            }
+                        }
+                    }
+                }
             }
 
-            // Ignore end whitespace
-            var lengthChanged = false;
-            for (; valueEnd >= valueStart; valueEnd--)
+            var valueEnd = headerLine.Length - 1;
+            // Ignore end whitespace. Generally there will no spaces
+            // so we will check the first before moving to a loop.
+            if (valueEnd > valueStart)
             {
                 var ch = headerLine[valueEnd];
-                if (ch != ByteTab && ch != ByteSpace)
+                // Do a fast check; as we now expect non-space, before moving into loop.
+                if (ch <= ByteSpace && (ch == ByteSpace || ch == ByteTab))
                 {
-                    break;
+                    // Is whitespace so move to loop
+                    valueEnd--;
+                    for (; valueEnd > valueStart; valueEnd--)
+                    {
+                        ch = headerLine[valueEnd];
+                        if (ch != ByteTab && ch != ByteSpace)
+                        {
+                            // Non-whitespace char found, valueEnd is now start of value.
+                            break;
+                        }
+                    }
                 }
-
-                lengthChanged = true;
             }
 
-            if (lengthChanged)
-            {
-                // Length changed
-                valueBuffer = new Span<byte>(headerLine + valueStart, valueEnd - valueStart + 1);
-            }
+            // Range end is exclusive, so add 1 to valueEnd
+            valueEnd++;
+            handler.OnHeader(name: headerLine[..nameEnd], value: headerLine[valueStart..valueEnd]);
 
-            var nameBuffer = new Span<byte>(headerLine, nameEnd);
+            return true;
 
-            handler.OnHeader(nameBuffer, valueBuffer);
+        Reject:
+            // Reject is a conditional jump forward as we expect most headers to be accepted,
+            // so we want it to be unpredicted by an unprimied branch predictor.
```

Comment:
How much are we really worried about the unprimed branch predictor case? Is returning false directly really that bad?

Reply:
Thought you'd pick up on that 😉

Doesn't look too bad

![image](https://user-images.githubusercontent.com/1142958/78941502-b901f480-7aaf-11ea-8d97-1a6c79a354c6.png)